tests/unit/lib/secrets/vault_csi_tests.py

```diff
@@ -0,0 +1,271 @@
+import datetime
+import json
+import shutil
+import string
+import tempfile
+import typing
+import unittest
+
+from pathlib import Path
+from unittest.mock import mock_open
+from unittest.mock import patch
+
+import gevent
+import pytest
+import typing_extensions
+
+from baseplate.lib.secrets import SecretNotFoundError
+from baseplate.lib.secrets import secrets_store_from_config
+from baseplate.lib.secrets import SecretsStore
+from baseplate.lib.secrets import VaultCSISecretsStore
+
+SecretType: typing_extensions.TypeAlias = typing.Dict[str, any]
+
+
+def write_secrets(secrets_data_path: Path, data: typing.Dict[str, SecretType]) -> None:
+    """Write secrets to the current data directory."""
+    for key, value in data.items():
+        secret_path = secrets_data_path.joinpath(key)
+        secret_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(secret_path, "w") as fp:
+            json.dump(value, fp)
+
+
+def write_symlinks(data_path: Path) -> None:
+    csi_path = data_path.parent
+    # This path can be monitored for changes
+    # https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/c697863c35d5431ec048b440d36550eb3ceb338f/pkg/util/fileutil/atomic_writer.go#L60-L62
+    data_link = Path(csi_path, "..data")
+    if data_link.exists():
+        # Simulate atomic update
+        new_data_link = Path(csi_path, "..data-new")
+        new_data_link.symlink_to(data_path)
+        new_data_link.rename(data_link)
+    else:
+        data_link.symlink_to(data_path)
+    human_path = Path(csi_path, "secret")
+    if not human_path.exists():
+        human_path.symlink_to(csi_path.joinpath("..data/secret"))
+
+
+def new_fake_csi(data: typing.Dict[str, SecretType]) -> Path:
+    """Creates a simulated CSI directory with data and symlinks.
+    Note that this would already be configured before the pod starts."""
+    csi_dir = Path(tempfile.mkdtemp())
```

Comment:
nit (optional): since we're using pytest as our test runner already, it might be nice to use pytest's `tmp_path` fixture rather than creating temporary directories manually: https://docs.pytest.org/en/latest/how-to/tmp_path.html

You wouldn't need to manually clean it up this way either, since pytest handles cleanup (by default it leaves the past couple test run outputs around which is helpful in case you want to inspect failures).

Reply:
While I would like to apply this suggestion, I _think_ this would require overhauling the test suite pretty significantly. So I'll omit this change, but TIL about this feature